### ISSUE TITLE
WaitFor command exit with 0

### ISSUE THREAD

**Issue by tisonkun** [testcontainers/testcontainers-rs#702]
Some image, like Postgres, would have a best match wait strategy like `pg_isready` returns 0.

Maybe we can implement such a wait strategy for testing with command?

**Comment by DDtKey** [comment]
Hi @tisonkun 👋 

Could you elaborate on this, please? 

In general, both [WaitFor](https://docs.rs/testcontainers/latest/testcontainers/core/wait/enum.WaitFor.html) and [CmdWaitFor](https://docs.rs/testcontainers/latest/testcontainers/core/enum.CmdWaitFor.html) supports waiting for exit code.

But I'm not sure I got the request correctly

**Comment by tisonkun** [comment]
> Also, Image implementation can override [exec_after_start](https://docs.rs/testcontainers/latest/testcontainers/core/trait.Image.html#method.exec_after_start), returning ExecCommands with their own CmdWaitFor (including exit code)

This seems the way I'd like to go. 

If `exec_after_start` can block `start` and thus `start` will finish after `exec_after_start` success (with retry), then it sounds a good solution.

**Comment by DDtKey** [comment]
Yes, these commands are executed before returning `Container` instance form the `start`

**Comment by tisonkun** [comment]
@DDtKey I try:

```
    fn exec_after_start(&self, _: ContainerState) -> Result<Vec<ExecCommand>, TestcontainersError> {
        Ok(vec![ExecCommand::new([
            "pg_isready",
            "-U",
            USERNAME,
            "-d",
            "postgres",
        ])
        .with_cmd_ready_condition(CmdWaitFor::exit_code(0))])
    }
```

But it panics with `called `Result::unwrap()` on an `Err` value: Exec(ExitCodeMismatch { expected: 0, actual: 2 })` insteadof retry until exit with 0.

**Comment by DDtKey** [comment]
Yes, as I mentioned above - there is no retry of the command itself for now. It seems to be a candidate for a separate feature.

But as a workaround you may consider to use `while` or `until` loops with bash for example 🤔

**Comment by DDtKey** [comment]
Btw, speaking of `postgres`, don't such conditions work for you? 

https://github.com/testcontainers/testcontainers-rs-modules-community/blob/66bbad597d4bbed30ef210e6a0afdb64089a3bb7/src/postgres/mod.rs#L86

These ones are used across different implementations of testcontainers (Java, Go, etc)

**Comment by ns-sjorgedeaguiar** [comment]
Here's my workaround to wait until a command finishes and printing errors when exist code is not 0:
```
let mut timeout = Duration::from_secs(5);
let mut exit_code = result.exit_code().await?;
loop {
    if exit_code.is_some() {
        break;
    }

    tokio::time::sleep(Duration::from_millis(100)).await;
    timeout -= Duration::from_millis(100);

    if timeout.as_millis() <= 0 {
        return Err("Timeout while waiting command to finish".into());
    }

    exit_code = result.exit_code().await?;
}

match exit_code {
    Some(0) => {}
    _ => {
        let code = result.exit_code().await?.unwrap_or(-1);
        let mut buffer = String::new();
        let mut stderr = result.stderr();
        stderr.read_to_string(&mut buffer).await?;
        buffer.split('\n').for_each(|line| {
            if !line.is_empty() {
                eprintln!("stderr: {}", line);
            }
        });
        return Err(format!("Failed to execute SQL (code = {})", code).into());
    }
};
```

**Comment by DDtKey** [comment]
I think ability to retry command should be incorporated into testcontainers. 
Let's keep this item, I'll implement it later if if no one willing to contribute appears (any help is appreciated since I have limited bandwidth and number of waiting features)

**Comment by AntoinePrv** [comment]
I came looking for this for Postgres as well. The message on stdout can do for me, but it would be nice to be able to use `pg_isready`.

For instance, testcontainers Node has:
```js
.withWaitStrategy(Wait.forSuccessfulCommand(`pg_isready -d ${POSTGRES_DB}`))
```

https://node.testcontainers.org/features/wait-strategies/#shell-command

**Comment by DDtKey** [comment]
I’m curious if #814 addresses the issue.

It’s slightly different, but it’s quite natural for Docker. It allows you to set a custom healthcheck (including a cmd, the number of retries, and so on) and utilize the healthcheck wait strategy.
